topo/node/ceos/ceos.go

```diff
@@ -30,24 +39,83 @@ func New(pb *topopb.Node) (node.Implementation, error) {
 }
 
 type Node struct {
-	pb *topopb.Node
+	pb      *topopb.Node
+	cliConn *scraplinetwork.Driver
 }
 
 func (n *Node) Proto() *topopb.Node {
 	return n.pb
 }
 
-var (
-	spawner = defaultSpawner
-)
+// WaitCLIReady attempts to open the transport channel towards a Network OS and perform scrapligo OnOpen actions
+// for a given platform. Retries indefinitely till success.
+func (n *Node) WaitCLIReady() error {
+	transportReady := false
+	for !transportReady {
+		if err := n.cliConn.Open(); err != nil {
+			log.Debugf("%s - Cli not ready - waiting.", n.pb.Name)
+			time.Sleep(time.Second * 2)
+			continue
+		}
+		transportReady = true
+		log.Debugf("%s - Cli ready.", n.pb.Name)
+	}
 
-func defaultSpawner(command string, timeout time.Duration, opts ...expect.Option) (expect.Expecter, <-chan error, error) {
-	return expect.Spawn(command, timeout, opts...)
+	return nil
 }
 
-var (
-	timeSecond = time.Second
-)
+// PatchCLIConnOpen sets the OpenCmd and ExecCmd of system transport to work with `kubectl exec` terminal.
+func (n *Node) PatchCLIConnOpen(ns string) error {
+	tIntf := reflect.ValueOf(n.cliConn.Transport)
```

Comment:
so this looks like you are just getting Transport that is part of the base.Driver?

// PatchCLIConnOpen sets the OpenCmd and ExecCmd of system transport to work with `kubectl exec` terminal.
func (n *Node) PatchCLIConnOpen(ns string) error {
	t, ok := n.cliConn.Driver.Transport.(*scraplitransport.System)
	if !ok {
		return fmt.Errorf("invalid underlying transport type: %T", n.cliConn.Driver.Transport)
	}
	t.ExecCmd = "kubectl"
	t.OpenCmd = []string{"exec", "-it", "-n", ns, n.pb.Name, "--", "Cli"}
	return nil
}

Reply:
> so this looks like you are just getting Transport that is part of the base.Driver?

Yeah -- scrapli has multiple transport options so we have to check the type... moreover we have to check for the *fields* because the mock test transport (that patches the reads/writes w/ the stored "session file") only embeds the transport type that we use here (`system` transport -- this is the subprocess wrapper one basically). Outside of the testing use case this could just type assert that `Transport` is of type `transport.System` but the testing bit adds a bit of fun to this.

So TL;DR we have to check if the transport has the `ExecCmd`/`OpenCmd` field attached to it.

Reply:
My 2c is that here the usage of reflect is somewhat innocent. 

We merely set the two fields of the underlying struct to make the testing work on the whole Generate SelfSigned function instead of the underlying scrapligo pieces

Reply:
Also tests are red because CI is picking not the tip of the branch for whatever reason

Reply:
you need to rebase this PR against head

Reply:
using reflect in production code should almost always be avoided...
If you need to access a lower struct member then - make the member public / provide an accessor 
in this case you just need a type assertion on an interface

Reply:
Done. Thank you